### PR TITLE
[ENH] Implementing ES-RNN

### PR DESCRIPTION
**Pull request by Ankit-1204**
#### Reference Issues/PRs
Mentioned at #6660 

#### What does this implement/fix? Explain your changes.
Implements the ES-RNN, with reference from https://www.sciencedirect.com/science/article/pii/S0169207019301153
for non, single and double seasonality

### CODE REVIEW

**Comment by Ankit-1204** [comment]
@fkiraly , Since this is my first time implementing a network for sktime, I could use some review related to integration,
- First off, I was comfortable with torch so I used that, I could change it if its not correct
- There may be some syntax error related to tensor, as I initially just used numpy(I'll fix that)

- [ESRNN](https://github.com/damitkwr/ESRNN-GPU/blob/master/es_rnn/model.py), I found this implementation, and I think my implementation for seasonality is constant and probably wrong, should I try to align it more with the given implementation?. This is the part I am confused about. I think the paper also mentioned the seasonality to evolve with time

**Review by fkiraly — DISMISSED** [review]
Great, this is nice!

Now, we should interface it as a forecaster - using the `forecaster_simple` extension template.

**Comment by geetu040** [comment]
> * [ESRNN](https://github.com/damitkwr/ESRNN-GPU/blob/master/es_rnn/model.py), I found this implementation from kaggle, and I think my implementation for seasonality is constant and probably wrong

can you link both the implementations you are referring?

> * First off, I was comfortable with torch so I used that, I could change it if its not correct

what does the other implementation use?

File: sktime/networks/es_rnn.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    # Deserializing the input
+    return (
+        level,
+        seasonality,
+        x / (level * seasonality[torch.arange(length) % season_length]),
+    )
+
+
+def _double_seasonal(alpha, beta, x):
+    pass
```
Comment (geetu040) [inline sktime/networks/es_rnn.py:69]: this is not used anywhere

File: sktime/networks/es_rnn.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        def __init__(
+            self,
+            input_shape,
+            hidden_size,
+            horizon,
+            num_layer,
+            alpha,
+            beta,
+            season_length,
+            seasonality="zero",
```
Comment (geetu040) [inline sktime/networks/es_rnn.py:112]: can you make these parameters more descriptive in the docstring, as well as better meaningful names would be appreciated for each parameter

File: sktime/networks/es_rnn.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        return self._ESRNN(
+            self.input_shape,
+            self.hidden_size,
+            self.horizon,
+            self.num_layer,
+            self.seasonality,
+        )
+
+    @classmethod
+    def get_test_params(cls, parameter_set="default"):
```
Comment (geetu040) [inline sktime/networks/es_rnn.py:196]: this should go in the forecasting class

**Review by geetu040 — CHANGES_REQUESTED** [review]
Thanks, this looks good so far.
- please add link on top for the file as a comment for the source code if you copied it from somewhere.
- see this estimator for reference: [SciNet Network Class](https://github.com/sktime/sktime/blob/main/sktime/networks/scinet.py), [Scinet Forecasting Class](https://github.com/sktime/sktime/blob/main/sktime/forecasting/scinet.py)

**Comment by Ankit-1204** [comment]
> can you link both the implementations you are referring?

oh I think I phrased it incorrectly, i meant i found [ESRNN](https://github.com/damitkwr/ESRNN-GPU/blob/master/es_rnn/model.py) mentioned on kaggle and compared that to what i wrote for this PR.

File: sktime/networks/es_rnn.py (hunk 1 above)
Comment (Ankit-1204) [reply sktime/networks/es_rnn.py:69]: yes, i was going to write that after _single_seasonal function was reviewed.

**Comment by geetu040** [comment]
> oh I think I phrased it incorrectly, i meant i found [ESRNN](https://github.com/damitkwr/ESRNN-GPU/blob/master/es_rnn/model.py) mentioned on kaggle and compared that to what i wrote for this PR.

I see, complete the forecasting class and we'll see how it works

File: sktime/networks/es_rnn.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        level = alpha * t + (1 - alpha) * level
+
+    return level, x / level
+
+
+def _single_seasonal(alpha, beta, season_length, x):
+    length = len(x)
+    # input should cover atleast one season length
+    if season_length > length:
+        season_length = length
```
Comment (VascoSch92) [inline sktime/networks/es_rnn.py:35]: This is just `season_lenght = max(season_lenght, lenght)`

Comment (Ankit-1204) [reply sktime/networks/es_rnn.py:35]: that's true, i did mean to add a warning for the user there regarding input window to be at least the size of season_length as suggested in the paper. Thanks for pointing that out.

File: sktime/networks/es_rnn.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+            Parameters
+            ----------
+            x : torch.Tensor
+                Input tensor of shape (batch_size, input_length).
+            """
+            if isinstance(x, np.ndarray):
+                x = torch.from_numpy(x.astype("float32"))
+            self._seasonality = self.seasonality
+            self._horizon = self.horizon
```
Comment (VascoSch92) [inline sktime/networks/es_rnn.py:141]: why are you defining these two new attributes?

File: sktime/networks/es_rnn.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            if self._seasonality == "zero":
+                level, new_x = _nonseasonal(x)
+
+                x_input = self.input_layer(new_x.float())
+                output = self.output_layer(x_input.unsqueeze(0))
+
+                output_leveled = output * level
+                return output_leveled
+            elif self.seasonality == "single":
```
Comment (VascoSch92) [inline sktime/networks/es_rnn.py:150]: what if self.seasonality is neither `zero` or `single`? The forward will return `None`. Is this the expected output?

Comment (Ankit-1204) [reply sktime/networks/es_rnn.py:150]: i left 'else', to cover the double seasonality situation. I hadn't implemented it at that point because i had some doubts regarding single_seasonality function i wrote which i wanted to cover first.

File: sktime/networks/es_rnn.py (hunk 5 above)
Comment (Ankit-1204) [reply sktime/networks/es_rnn.py:141]: right, i remember adding that at that point assuming i might have to modify there values, but this was mostly boiler plate, i have removed it while working on the forecasting class. I'll try push that as soon as i get the chance.

**Comment by Ankit-1204** [comment]
@fkiraly @geetu040 can you review it now?

**Comment by fkiraly** [comment]
I started the full suite of tests

File: sktime/networks/es_rnn.py (hunk 2 above)
Comment (Ankit-1204) [reply sktime/networks/es_rnn.py]: yes, its done now.

File: sktime/forecasting/esrnn.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+from sktime.utils.dependencies import _check_soft_dependencies
+
+
+class ESRNNForecaster(BaseDeepNetworkPyTorch):
+    r"""
+    Exponential Smoothing Recurrant Neural Network.
+
+    This model combines Exponential Smoothing (ES) and (LSTM) networks
+    for time series forecasting. ES is used to balance the level and
+    seasonality of the series.
```
Comment (geetu040) [inline sktime/forecasting/esrnn.py:17]:
```suggestion
    seasonality of the series. This method has been
    proposed in [1]_.
```

File: sktime/forecasting/esrnn.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+    References
+    ----------
+    [1] Smyl, S. (2020). A hybrid method of exponential smoothing and
+        recurrent neural networks for time series forecasting.
+
+        https://www.sciencedirect.com/science/article/pii/S0169207019301153
```
Comment (geetu040) [inline sktime/forecasting/esrnn.py:24]: 3 comments here:
1. "References" should be after "Parameters", just a normal standard.
2. Make sure this reference is used in the above description of model
3. I am not sure if this will be rendered correctly, please see how the reference is written in [LTSFLinearForecaster](https://github.com/sktime/sktime/blob/main/sktime/forecasting/ltsf.py#L47)

File: sktime/forecasting/esrnn.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        if self.criterion:
+            if self.criterion in self.loss_list:
+                return self.loss_list[self.criterion]()
+            else:
+                loss = ESRNN().DefaultLoss()
+                return loss
+        else:
+            # default criterion
+            loss = ESRNN().DefaultLoss()
+            return loss
```
Comment (geetu040) [inline sktime/forecasting/esrnn.py:143]:
```suggestion
    def _instantiate_criterion(self):
        if self.criterion:
            return super()._instantiate_criterion()
        return ESRNN.pinball_loss()
```
also see the comments on loss in network class

File: sktime/networks/es_rnn.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+                        * seasonality_2[:, -self.horizon :, :]
+                    )
+                    return output_leveled
+
+        self._network_class = _ESRNN
+        self.loss = PinballLoss
+
+    def DefaultLoss(self):
+        """Return the default Pinball Loss."""
+        return self.loss()
```
Comment (geetu040) [inline sktime/networks/es_rnn.py:322]:
```suggestion
    def pinball_loss(self):
        """Return the default Pinball Loss."""
        return self.loss()
```

File: sktime/networks/es_rnn.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+
+    def build_network(self):
```
Comment (geetu040) [inline sktime/networks/es_rnn.py:324]:
```suggestion
    def _build(self):
```
following the convention of other pytorch models

File: sktime/networks/es_rnn.py (hunk 3 above)
Comment (geetu040) [reply sktime/networks/es_rnn.py:337]: this still remains resolved

**Review by geetu040 — CHANGES_REQUESTED** [review]
The core algorithmic logic looks good, just the interface that needs improvements.

Please make the following changes and see the comments
- keep the networks and forecaster file names consistent, choose either of `es_rnn` or `esrnn` for both.
- add the model to docs file: [docs/source/api_reference/forecasting.rst](https://github.com/sktime/sktime/blob/main/docs/source/api_reference/forecasting.rst)
- I think you don't need to overload `_fit` and `_predict`, try to make it work with a custom dataset class and overload `build_pytorch_train_dataloader` and `build_pytorch_pred_dataloader` instead if necessary.

Let me know if you need any help.

**Comment by geetu040** [comment]
And sorry for a belated reply, just ping when you are done or have any question

**Comment by Ankit-1204** [comment]
@geetu040 can you review it again now? Thanks.

File: docs/source/api_reference/forecasting.rst [hunk 12]
```diff
@@ -292,6 +292,14 @@ Exponential smoothing based forecasters
...
     StatsForecastAutoTheta
 
+.. currentmodule:: sktime.forecasting.es_rnn
+
+.. autosummary::
+    :toctree: auto_generated/
+    :template: class.rst
+
+    ESRNNForecaster
+
```
Comment (geetu040) [inline docs/source/api_reference/forecasting.rst:302]: move this under the heading "Deep learning based forecasters"

File: sktime/forecasting/es_rnn.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+    Exponential Smoothing Recurrant Neural Network.
+
+    This model combines Exponential Smoothing (ES) and (LSTM) networks
+    for time series forecasting. ES is used to balance the level and
+    seasonality of the series. This method has been proposed in [1]_.
+
+    Parameters
+    ----------
+    input_shape : int
+        Number of features in the input
```
Comment (geetu040) [inline sktime/forecasting/es_rnn.py:89]:
```suggestion
```
not needed, this is inferred from data, fix at other places as well

File: sktime/forecasting/es_rnn.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+    criterion_kwargs : dict, default=None
+        keyword arguments to pass to criterion
+    optimizer : torch.optim.Optimizer, default=torch.optim.Adam
+        optimizer to be used for training
+    optimizer_kwargs : dict, default=None
+        keyword arguments to pass to optimizer
+    window : int
+        Size of Input window, default=5
+    lr_rate : int
+        Learning rate for training
```
Comment (geetu040) [inline sktime/forecasting/es_rnn.py:117]:
```suggestion
    lr : int
        Learning rate for training
```
should be simply called `lr` (learning rate) instead, fix at all places

File: sktime/forecasting/es_rnn.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+    https://www.sciencedirect.com/science/article/pii/S0169207019301153
+
+    Examples
+    --------
+    >>> from sktime.forecasting.es_rnn import ESRNNForecaster # doctest: +SKIP
+    >>> from sktime.datasets import load_airline
+    >>> y = load_airline()
+    >>> forecaster = ESRNNForecaster() # doctest: +SKIP
+    >>> forecaster.fit(y, fh=[1,2,3]) # doctest: +SKIP
+    >>> y_pred = forecaster.predict() # doctest: +SKIP
```
Comment (geetu040) [inline sktime/forecasting/es_rnn.py:133]: I have tried this example with different parameters to `ESRNNForecaster` but they don't produce any good results on `load_airline`, can you try to find an example with such parameters that result in good forecasts?

**Review by geetu040 — CHANGES_REQUESTED** [review]
Thanks for updating, looks good to go, just some minor changes requested

File: sktime/forecasting/es_rnn.py (hunk 15 above)
Comment (Ankit-1204) [reply sktime/forecasting/es_rnn.py]: I have updated the example with parameters which were giving better forecast, mainly involved increasing epoch, layers and lr. Please take a look.

Comment (geetu040) [reply sktime/forecasting/es_rnn.py]: This improves the forecast, but are many epochs. Can you try some hyperparameters that don't take this long to train? You may try scaling the data, choosing a different learning rate.

Comment (Ankit-1204) [reply sktime/forecasting/es_rnn.py]: Normalized the data and increased the lr, now it gives better forecasts with reduced epochs

**Review by geetu040 — DISMISSED** [review]
LGTM!

**Comment by geetu040** [comment]
@Ankit-1204 can you merge main branch and resolve the conflicts?

**Review by geetu040 — APPROVED** [review]
Lets wait for the tests to finish.
